Serial.md

```diff
@@ -0,0 +1,312 @@
+
+Connections will be
+```
+MISO -> RX
+SCK -> TX
+GND -> GND
+```
+
+(Store this file in your esphome configuration directory, for example `megadesk.h`)
+```
+#include "esphome.h"
+
+class Megadesk : public Component, public Sensor, public UARTDevice {
+ public:
+  Megadesk(UARTComponent *parent) : UARTDevice(parent) {}
+
+  void setup() override {}
+
+  int digits=0;
+
+  int readdigits()
+  {
+    int r;
+    while ((r = read()) > 0) {
+      if ((r < 0x30) || (r > 0x39)) {
+        // non-digit we're done, return what we have
+        return digits;
+      }
+      // it's a digit, add with base10 shift
+      digits = 10*digits + (r-0x30);
+      // keep reading...
+    }
+    return -1;
+  }
+
+  void recvData()
+  {
+    const int numChars = 2;
+    const int numFields = 4; // read/store all 4 fields for simplicity, use only the last 3.
+    // static variables allows segmented/char-at-a-time decodes
+    static uint16_t receivedBytes[numFields];
+    static uint8_t ndx = 0;
+    int r; // read char/digit
+
+    // read 2 chars
+    while ((ndx < numChars) && ((r = read()) != -1))
+    {
+      if ((ndx == 0) && (r != '>'))
+      {
+        // first char is not Tx, keep reading...
+        continue;
+      }
+      receivedBytes[ndx] = r;
+      ++ndx;
+    }
+    // read ascii digits
+    while ((ndx >= numChars) && ((r = readdigits()) != -1)) {
+      receivedBytes[ndx] = r;
+      digits = 0; // clear
+      if (++ndx == numFields) {
+        // thats all 4 fields. parse/process them now and break-out.
+        parseData(receivedBytes[1],
+                  receivedBytes[2],
+                  receivedBytes[3]);
+        ndx = 0;
+        return;
+      }
+    }
+  }
+
+  void parseData(byte command, uint16_t position, uint8_t push_addr)
+  {
+    if (command == '=')
+    {
+      publish_state(position);
+    }
+  }
+
+  void loop() override {
+    while (available()) {
+      recvData();
+    }
+  }
+};
+```
+
+
+esphome YAML
+ ```
+esphome:
+  name: megadesk
+  platform: ESP8266
+  board: d1_mini
+  includes:
+    - megadesk.h
+  on_boot:
+    priority: -100
+    then:
+      - uart.write: "<C0.0."
+
+logger:
+  baud_rate:0
+
+api:
+  password: ""
+
+ota:
+  password: ""
+
+wifi:
+  ssid: "xxxxxxxxxxx"
+  password: "xxxxxxxxx"
+
+  ap:
+    ssid: "Desk Fallback Hotspot"
+    password: "xxxxxxxxxx"
+
+captive_portal:
+
+web_server:
+  port: 80
+
+uart:
+  id: uart_desk
+  baud_rate: 115200
+  tx_pin: D0
+  rx_pin: D1
+
+sensor:
+- platform: custom
+  lambda: |-
+    auto megadesk = new Megadesk(id(uart_desk));
+    App.register_component(megadesk);
+    return { megadesk };
+  sensors:
+  - id: megadesk_raw
+    internal: true
+    on_value:
+      then:
+        - component.update: megadesk_height_inches
+        - component.update: megadesk_height_cm
+        - component.update: megadesk_height_raw
```

Comment:
cm/inch heights - nice!

Reply:
Heights have NOT been verified to be accurate.  The min and max are based upon the heights stated by Ikea.

Reply:
Would it be possible for you measure what the true minimum and maximum heights are in either centimeters or inches?  My desk is on carpet, which would cause height inaccuracies.